src/states/abstract/State.js

```diff
@@ -0,0 +1,126 @@
+import videojs from 'video.js';
+
+export default class State {
+
+  constructor(player) {
+    this.player = player;
+  }
+
+  /*
+   * This is the only allowed way to perform state transitions. State transitions usually
+   * happen in player event handlers. They can also happen recursively in `init`. They
+   * should _not_ happen in `cleanup`.
+   */
+  transitionTo(NewState, ...args) {
+    const player = this.player;
+    const previousState = this;
+
+    previousState.cleanup();
+    const newState = new NewState(player);
+
+    player.ads._state = newState;
+    player.ads.debug(previousState.constructor.name + ' -> ' + newState.constructor.name);
+    newState.init(player, ...args);
+  }
+
+  /*
+   * Implemented by subclasses to provide initialization logic when transitioning
+   * to a new state.
+   */
+  init() {}
+
+  /*
+   * Implemented by subclasses to provide cleanup logic when transitioning
+   * to a new state.
+   */
+  cleanup() {}
+
+  /*
+   * Default event handlers. Different states can override these to provide behaviors.
+   */
+  onPlay() {}
+  onPlaying() {}
+  onEnded() {}
+  onAdsReady() {
+    videojs.log.warn('Unexpected adsready event');
+  }
+  onAdsError() {}
+  onAdsCanceled() {}
+  onAdTimeout() {}
+  onAdStarted() {}
+  onContentUpdate() {}
+  onContentResumed() {}
+  onContentEnded() {
+    videojs.log.warn('Unexpected contentended event');
+  }
+  onNoPreroll() {}
+  onNoPostroll() {}
+
+  /*
+   * Method handlers. Different states can override these to provide behaviors.
+   */
+  startLinearAdMode() {
+    videojs.log.warn('Unexpected startLinearAdMode invocation ' +
+      '(State via ' + this.constructor.name + ')');
+  }
+  endLinearAdMode() {
+    videojs.log.warn('Unexpected endLinearAdMode invocation ' +
+      '(State via ' + this.constructor.name + ')');
+  }
+  skipLinearAdMode() {
+    videojs.log.warn('Unexpected skipLinearAdMode invocation ' +
+      '(State via ' + this.constructor.name + ')');
+  }
+
+  /*
+   * Overridden by ContentState and AdState. Should not be overriden elsewhere.
+   */
+  isAdState() {}
```

Comment:
should this return a default boolean? Seems incorrect to not have anything returned

Reply:
If anything, I think the default should be to throw an error. My intention was to make it clear that it must be overwritten.

Reply:
I like the idea of throwing an error

Reply:
Doing that